content/tutorial-plotting-fractals.md

```diff
@@ -0,0 +1,558 @@
+
+```{code-cell} ipython3
+def general_mandelbrot(mesh,c=-1,f=np.square,num_iter=100,radius=2):
+    
+    z = mesh.copy()
+    diverge_len = np.zeros(z.shape) 
+        
+    for i in range(num_iter):
+        z[np.abs(z)< radius] = f(z[np.abs(z)< radius]) + c
+        diverge_len[np.abs(z)< radius] += 1
+        
+    return diverge_len
+```
+
+One cool set of fractals that can be plotted using our general Mandelbrot function are ones of the form $f(z) = z^n + c$ for some positive integer n. A very cool pattern which emerges is that the number of regions that 'stick out' matches the degree in which we raise the function to while iterating over it.
+
+```{code-cell} ipython3
+fig, axes = plt.subplots(2,3,figsize=(8,8))
+degree = 2
+
+for i in range(2):
+    for j in range(3):
+        
+        power = lambda z: np.power(z,degree) #Create power function for current degree 
+
+        diverge_len = general_mandelbrot(mesh,f=power,num_iter=15)
+        axes[i,j].imshow(diverge_len,extent=[-2,2,-2,2],cmap='binary')
+        axes[i,j].set_title(f'$f(z) = z^{degree} -1$')
+                
+        degree += 1
+        
+fig.tight_layout();
+```
+
+Needless to say, there is a large amount of exploring that can be done by fiddling with the inputted function, value of c, number of iterations, radius and even the density of the mesh and choice of colours.
+
++++
+
+## Newton Fractals
+
+Newton fractals are a specific class of fractals, where iterations involve adding or subtracting the ratio of a function (often a polynomial) and its derivative to the input values. Mathematically, it can be expressed as:
+
+$z := z - \frac{f(z)}{f'(z)}$
+
+We will define a general version of the fractal which will allow for different variations to be plotted by passing in our functions of choice.
+
+```{code-cell} ipython3
+def newton_fractal(mesh,f,df,num_iter=10,r=2):
+    
+    z = mesh.copy()
+    diverge_len = np.zeros(z.shape) 
+        
+    for i in range(num_iter):
+        pz = f(z[np.abs(z)<r])
+        dp = df(z[np.abs(z)<r])
+        z[np.abs(z)<r] = z[np.abs(z)<r] - pz/dp
+        diverge_len[np.abs(z)<r] += 1
+        
+    return diverge_len
+```
+
+Now we can experiment with some different polynomials.
+
+For $p(z) = z^8 + 15 z^4 - 16$, we have 
+
+
+$\frac{dp}{dz} = 8z^7 + 60z^3$.
+
+Putting the polynomials to code and calling the function gives us:
+
+```{code-cell} ipython3
+def poly(z):
+    return np.power(z,8) + 15* np.power(z,4) -16
+
+def derivative(z):
+    return 8* np.power(z,7) + 60 * np.power(z,3)
+```
+
+```{code-cell} ipython3
+output = newton_fractal(mesh,poly,derivative,num_iter=15,r=2)
+kwargs = {'title':'f(z) = z - \dfrac{(z^8 + 15z^4 - 16)}{(8z^7 + 60z^3)}', 'cmap':'copper'}
+
+plot_fractal(output,**kwargs)
+```
```

Comment:
This is a nice opportunity to highlight [NumPy's polynomial package](https://numpy.org/doc/stable/reference/routines.polynomials.html). Note that polynomial objects should render w/ TeX in Jupyter notebooks.

````suggestion
Now we can experiment with some different polynomials.
For example, let's try a higher-degree polynomial:

```{code-cell} ipython3
p = np.polynomial.Polynomial([-16, 0, 0, 0, 15, 0, 0, 0, 1])
p
```

which has the derivative:

```{code-cell} ipython3
p.deriv()
```

```{code-cell} ipython3
output = newton_fractal(mesh,p ,p.deriv(), num_iter=15,r=2)
kwargs = {'title':'f(z) = z - \dfrac{(z^8 + 15z^4 - 16)}{(8z^7 + 60z^3)}', 'cmap':'copper'}

plot_fractal(output,**kwargs)
```
````